### PR TITLE
Predeploy token fee and account contracts

### LINKED ISSUE #55: A list of pre-funded test accounts
## Feature Request

To have a list of pre-populated accounts the same as ganache of Hardhat node does.

Bonus if we can have a single key, like HD-Path style, be the source of generating all of them.

## Questions

Is this possible to do?
Is it wise to do?

## Disclaimer

I am still getting used to these ZK rollup style ecosystems.
My perspective is a builder who needs to make tooling for Starknet and needs it to be roughly compatible / similar to other existing Python Ethereum tooling.

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- Introduces token for charging fees (ERC20)
- Predeploys accounts
- Introduces new CLI options for account manipulation
- Closes #55

## Development related changes

- Introduce `psutil` dev dependency for postman testing (for killing of children processes)

## Checklist:

- [x] Performed a self-review of the code
- [x] Rebased to the base branch
- [x] Documented the changes
- [x] Updated the tests
- [x] All tests are passing

### CODE REVIEW

File: starknet_devnet/util.py [hunk 1]
```diff
@@ -124,10 +165,35 @@ def parse_args():
...
         help="Disables deploy tx hash calculation"
     )
+    parser.add_argument(
+        "--accounts",
+        type=int,
+        help=f"Specify the number of accounts to be predeployed; defaults to {DEFAULT_ACCOUNTS}",
+        default=DEFAULT_ACCOUNTS
+    )
+    parser.add_argument(
+        "--initial-balance", "-e",
```
Comment (badurinantun) [inline starknet_devnet/util.py:175]: Why `-e`?

File: starknet_devnet/util.py [hunk 2]
```diff
@@ -61,6 +80,28 @@ def parse_dump_on(option: str):
         return DumpOn[option.upper()]
     sys.exit(f"Error: Invalid --dump-on option: {option}. Valid options: {DUMP_ON_OPTIONS_STRINGIFIED}")
 
+def int_or_float(value: str):
+    """Parses the value as either an int or a float."""
```
Comment (badurinantun) [inline starknet_devnet/util.py:84]: But it always return an int, please document that. Perhaps change name to `numberish` or something

**Review by badurinantun — CHANGES_REQUESTED** [review]
Looks good, no major requests

File: starknet_devnet/state.py [hunk 3]
```diff
@@ -18,13 +23,46 @@ def __set_starknet_wrapper(self, starknet_wrapper: StarknetWrapper):
...
     def load(self, load_path: str):
         """Loads starknet wrapper from path"""
         self.__set_starknet_wrapper(StarknetWrapper.load(load_path))
 
+    def generate_accounts(self, n_accounts: int, initial_balance: int, seed: int):
+        """Generates accounts without deploying them"""
+        random_generator = random.Random()
+
+        if seed is None:
+            seed = random_generator.getrandbits(32)
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/state.py:41]: Ganache doesn't have a default seed. I think that's usually the case with tools that have a seed option: each time they generate something different unless a seed is specified.

File: starknet_devnet/util.py [hunk 4]
```diff
@@ -124,10 +165,35 @@ def parse_args():
...
+        help="Specify the seed for randomness of accounts to be predeployed"
+    )
     parser.add_argument(
         "--start-time",
         action=NonNegativeAction,
-        help="Specify the start time of the genesis block in Unix time"
+        help="Specify the start time of the genesis block in Unix time seconds"
+    )
+    parser.add_argument(
+        "--gas-price", "-g",
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/util.py:192]: Was following this: https://github.com/trufflesuite/ganache

File: starknet_devnet/util.py (hunk 1 above)
Comment (FabijanC) [reply starknet_devnet/util.py:175]: Was following this: https://github.com/trufflesuite/ganache

File: starknet_devnet/starknet_wrapper.py [hunk 5]
```diff
@@ -57,13 +59,28 @@ def __init__(self, config: DevnetConfig):
...
     @staticmethod
     def load(path: str) -> "StarknetWrapper":
         """Load a serialized instance of this class from `path`."""
         with open(path, "rb") as file:
             return pickle.load(file)
 
+    async def initialize(self):
+        """Initialize the underlying starknet instance, fee_token and accounts."""
+        if not self.__initialized:
+            starknet = await self.__get_starknet()
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/starknet_wrapper.py:76]: Good idea, I'll deal with this in a separate PR.

File: starknet_devnet/fee_token.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+from starknet_devnet.util import Uint256
+
+class FeeToken:
+    """Wrapper of token for charging fees."""
+
+    DEFINITION = ContractDefinition.load(load_nearby_contract("ERC20"))
+    # HASH = to_bytes(compute_contract_hash(contract_definition=DEFINITION))
+    HASH = 375899817338126263298463755162657787890597705735749339531748983767835688120
+    HASH_BYTES = to_bytes(HASH)
+    SALT = 10
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/fee_token.py:25]: https://github.com/Shard-Labs/starknet-devnet/pull/115/commits/1750b02f857613e9d23de4d21186c84bd9ea0cba

File: starknet_devnet/account.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+class Account:
+    """Account contract wrapper."""
+
+    CONTRACT_PATH = "accounts_artifacts/OpenZeppelin/0.1.0/Account.cairo/Account"
+    DEFINITION = ContractDefinition.load(load_nearby_contract(CONTRACT_PATH))
+    # HASH = compute_contract_hash(contract_definition=DEFINITION))
+    HASH = 361479646297615797917493841430922492724680358320444679508058603177506550951
+    HASH_BYTES = to_bytes(HASH)
+    SALT = 20
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/account.py:30]: https://github.com/Shard-Labs/starknet-devnet/pull/115/commits/1750b02f857613e9d23de4d21186c84bd9ea0cba

File: starknet_devnet/account.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+from starknet_devnet.util import Uint256
+
+class Account:
+    """Account contract wrapper."""
+
+    CONTRACT_PATH = "accounts_artifacts/OpenZeppelin/0.1.0/Account.cairo/Account"
+    DEFINITION = ContractDefinition.load(load_nearby_contract(CONTRACT_PATH))
+    # HASH = compute_contract_hash(contract_definition=DEFINITION))
+    HASH = 361479646297615797917493841430922492724680358320444679508058603177506550951
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/account.py:26]: https://github.com/Shard-Labs/starknet-devnet/pull/115/commits/1750b02f857613e9d23de4d21186c84bd9ea0cba

File: starknet_devnet/server.py [hunk 9]
```diff
@@ -53,12 +67,32 @@ def main():
...
+    # Uncomment this once fork support is added
+    # origin = Origin(args.fork) if args.fork else NullOrigin()
+    # starknet_wrapper.origin = origin
+
+    handle_accounts(args)
+    handle_load(args)
+    handle_dump(args)
+    handle_lite_mode(args)
+    handle_start_time(args)
+    handle_gas_price(args)
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/server.py]: https://github.com/Shard-Labs/starknet-devnet/pull/115/commits/cf9335c02d5e145deb15c25a7f62ef16ae0a2c76

File: starknet_devnet/util.py [hunk 10]
```diff
@@ -61,6 +80,28 @@ def parse_dump_on(option: str):
...
+            warnings.warn(f"Warning: {value} lost precision due to rounding!")
+        return int_value
+    except ValueError:
+        pass
+
+    sys.exit(f"Error: {value} is neither an integer nor a float")
+
+DEFAULT_ACCOUNTS = 10
+DEFAULT_INITIAL_BALANCE = 10 ** 21
+DEFAULT_GAS_PRICE = 10 ** 11
```
[the remark this answers is not in the document]
Comment (FabijanC) [reply starknet_devnet/util.py]: https://github.com/Shard-Labs/starknet-devnet/pull/115/commits/0a6dba129c9db9bf305ae2e623db1ccce3a82d9a

File: starknet_devnet/blueprints/base.py [hunk 11]
```diff
@@ -64,3 +65,14 @@ def set_time():
     state.starknet_wrapper.set_block_time(time_s)
 
     return jsonify({"next_block_timestamp": time_s})
+
+@base.route("/get_account_balance", methods=["GET"])
```
[the remark this answers is not in the document]
Comment (badurinantun) [reply starknet_devnet/blueprints/base.py]: Perhaps they should know the ABI

**Review by badurinantun — APPROVED** [review]

**Pull request merged**